components/suggest/src/db.rs

```diff
@@ -197,58 +206,106 @@ impl<'a> SuggestDao<'a> {
+        let suggestions = self.conn.query_rows_and_then_cached(
+            "SELECT s.id, k.rank, s.title, s.url
+             FROM suggestions s
+             JOIN keywords k ON k.suggestion_id = s.id
+             WHERE s.provider = :provider AND
+             k.keyword = :keyword",
+            named_params! {
+                ":keyword": keyword_lowercased,
+                ":provider": SuggestionProvider::Wikipedia
+            },
+            |row| -> Result<Suggestion> {
+                let suggestion_id: i64 = row.get("id")?;
+                let title = row.get("title")?;
+                let raw_url = row.get::<_, String>("url")?;
+
+                let keywords: Vec<String> = self.conn.query_rows_and_then_cached(
+                    "SELECT keyword FROM keywords
+                     WHERE suggestion_id = :suggestion_id AND rank >= :rank
+                     ORDER BY rank ASC",
+                    named_params! {
+                        ":suggestion_id": suggestion_id,
+                        ":rank": row.get::<_, i64>("rank")?,
+                    },
+                    |row| row.get(0),
+                )?;
+                let icon = self.conn.try_query_one(
+                    "SELECT i.data
+                     FROM icons i
+                     JOIN wikipedia_custom_details s ON s.icon_id = i.id
+                     WHERE s.suggestion_id = :suggestion_id",
+                    named_params! {
+                        ":suggestion_id": suggestion_id
+                    },
+                    true,
+                )?;
+                Ok(Suggestion::Wikipedia {
+                    title,
+                    url: raw_url,
+                    full_keyword: full_keyword(keyword_lowercased, &keywords),
+                    icon,
+                })
+            },
+        )?;
+        Ok(suggestions)
+    }
+
+    pub fn fetch_amo_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
+        let keyword_lowercased = &query.keyword.to_lowercase();
+        let (keyword_prefix, keyword_suffix) = split_keyword(keyword_lowercased);
+        let suggestions = self.conn.query_rows_and_then_cached(
+            "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, k.confidence, k.keyword_suffix
+                     FROM suggestions s
+                     JOIN prefix_keywords k ON k.suggestion_id = s.id
+                     WHERE k.keyword_prefix = :keyword_prefix AND s.provider = :provider",
+             named_params! {
+                ":keyword_prefix": keyword_prefix,
+                ":provider": SuggestionProvider::Amo
+            },
+            |row| -> Result<Option<Suggestion>>{
+                let suggestion_id: i64 = row.get("id")?;
+                let title = row.get("title")?;
+                let raw_url = row.get::<_, String>("url")?;
+                let score = row.get::<_, f64>("score")?;
+
+                let full_suffix = row.get::<_, String>("keyword_suffix")?;
```

Comment:
While you're here, do you think it's worth lifting the `full_suffix.starts_with(keyword_suffix)` check up here—so that we don't even query `amo_custom_details` if the suffixes don't match?

In other words:

```rs
let full_suffix = row.get::<_, String>("keyword_suffix")?;
if !full_suffix.starts_with(keyword_suffix) {
    return Ok(None);
}
self.conn.query_row_and_then(
    "SELECT amo.description, amo.guid, amo.rating, amo.icon_url, amo.number_of_ratings
```

Reply:
Ooooh, I like the `.then(|| { ... }).transponse()`! ✨